### PR TITLE
feat: Switch to the v2 Twitter API

### PR DESCRIPTION
**Pull request by MattIPv4**
breaking change: Moves to using the v2 API. Twitter apps need to be in a project to access the v2 API.

breaking change: When retweeting, the only URL returned now is the original tweet, not the retweeter's tweet.

Replaces the `twitter` library with `twitter-api-v2`, and switches the logic over to using the v2 Twitter API where possible.

This removes the need for the Ads API as the v2 tweet endpoint supports polls directly. This still uses the v1 API for media, but essentials API access will allow access to this as there is no v2 equivalent currently.

### CODE REVIEW

File: lib/common/tweet.js [hunk 1]
```diff
@@ -65,139 +65,34 @@ async function handleTweet(client, tweet, name) {
...
-  //       (error) => {
-  //         if (error) return reject(error);
-  //         resolve();
-  //       }
-  //     );
-  //   });
-  // }
-
-  return id;
+  const mediaId = await client.v1.uploadMedia(file, { mimeType: mime.lookup(file) });
```
Comment (IstoraMandiri) [inline lib/common/tweet.js:69]: Does this mean that the v2 API does not support media uploads?

Will twitter accounts without v1 access be unable to upload media?

Comment (MattIPv4) [reply lib/common/tweet.js:69]: As I note in the description above:

> This still uses the v1 API for media, but essentials API access will allow access to this as there is no v2 equivalent currently

**Comment by MattIPv4** [comment]
Only remaining work here now is to update the README and docs directory to remove references to the Ads API, to remove guidance on requesting elevated access as only essentials access is required now (v2 + v1 media), and to remove the comment that alt text cannot be set as it now can be.

**Comment by IstoraMandiri** [comment]
I have extensively tested each tweet type in production of this branch over at  https://github.com/IstoraMandiri/twitter-together-testing and https://twitter.com/testing_tt_

Everything seems to be working other than some minor preview-related bugs and misc improvements, which I have added issues to this repo for.

**Review by gr2m — CHANGES_REQUESTED** [review]
great work!!

One more thing we need to change is the release.yml workflow here:
https://github.com/twitter-together/action/blob/dffa59cb5c1b16fd6aca5014c6c30cfc4e04d23f/.github/workflows/release.yml#L23

This won't be necessary once I get to work on https://github.com/twitter-together/action/issues/205 😢

**Comment by MattIPv4** [comment]
Should I bump package.json to 3.0.0 at the same time, or should I leave that for the bot to handle still?

**Comment by gr2m** [comment]
> Should I bump package.json to 3.0.0 at the same time, or should I leave that for the bot to handle still?

that should be done by semantic-release via this configuration:
https://github.com/twitter-together/action/blob/dffa59cb5c1b16fd6aca5014c6c30cfc4e04d23f/package.json#L60-L67

**Comment by MattIPv4** [comment]
Cool, will bump the workflow, pls hold

**Review by gr2m — APPROVED** [review]
# 🚀:shipit:

**Comment by IstoraMandiri** [comment]
Nice, can confirm the Twitter App creation docs are also looking very good and match my recent experience.

**Comment by MattIPv4** [comment]
@gr2m Is there anything blocking merging this? Wasn't sure if I can boop the button, or if you'd left it unmerged for a reason?

**Comment by MattIPv4** [comment]
:sus: Why did it release it as minor

**Comment by MattIPv4** [comment]
Gregor, can you do some magic to unrelease that minor release, 'cause this is very likely a breaking change for some folks 🤔

**Comment by IstoraMandiri** [comment]
Pretty sure @MattIPv4 is right, @gr2m, as presumably current v1 API users would need to set up new keys to use the v2 endpoints?  

I cannot easily test this unfortunately.

**Comment by MattIPv4** [comment]
_Most_ users should be safe, it's only breaking for those that're using apps outside a project, as legacy apps only have v1 access.

I'm guessing the bot didn't look at the PR body, only the commit body (which just had the normal commit log w/o the breaking change warnings). Automation 😓

**Comment by IstoraMandiri** [comment]
I notice that the build is also built under 

```
/***/ 4147:
/***/ ((module) => {

"use strict";
module.exports = {"i8":"2.1.2"};
```

Perhaps bump the package.json to 3.0.0 and rebuild?

**Comment by gr2m** [comment]
> :sus: Why did it release it as minor

Ouch, sorry :(

semantic-release needs a `BREAKING CHANGE: ...` in the commit body in order to do a breaking release. 
I've documented the conventions here: https://github.com/twitter-together/action/blob/main/CONTRIBUTING.md#merging-the-pull-request--releasing-a-new-version

As we don't release to npm, we can just delete the git tag/github release and have semantic-release do a breaking release with an empty commit. I'm on it

**Comment by gr2m** [comment]
the above comment makes no sense to me 🤷🏼 

Here are the two relevant releases

- https://github.com/twitter-together/action/releases/tag/v2.2.1
- https://github.com/twitter-together/action/releases/tag/v3.0.0

2.x maintenance releases can be done with pull requsets against [the `2.x` branch](https://github.com/twitter-together/action/tree/2.x)

**Comment by MattIPv4** [comment]
Ty for the fix -- yah, figured it was probably the lack of breaking change mention in the commit. I'd read the doc before but I think parsed it as needing breaking change in the PR body, not commit body (though commit makes far, far more sense).

**Comment by gr2m** [comment]
semantic-release only works on `git`. The interaction with GitHub is just a plugin. Hence the need to have correct commit messages. I'd love a better integration with GitHub using an app for a long time now. The day just need more hours 🤣

**Comment by MattIPv4** [comment]
Makes total sense, I just had a lapse in brain function 😂

**Comment by MattIPv4** [comment]
To Istora's point above though, https://github.com/twitter-together/action/blob/v3/dist/index.js does declare itself as v2 still though 'cause it was built before the package.json bump was done.

Is there a semantic-release plugin that can run the build command after the version bump, and then commit both the updated package.json and dist.js? (Which I think is really all that's needed to solve for #205)

**Comment by gr2m** [comment]
I think this can be done via https://github.com/semantic-release/git

**Pull request merged**